adapters/DuAd/DuAdAdapter/GADDuAdNativeAd.m

```diff
@@ -0,0 +1,354 @@
+}
+
+- (NSDecimalNumber *)starRating {
+    return nil;
+}
+
+- (NSString *)store {
+    return nil;
+}
+
+- (NSString *)price {
+    return nil;
+}
+
+#pragma mark - GADMediatedNativeAdDelegate
+
+- (void)mediatedNativeAd:(id<GADMediatedNativeAd>)mediatedNativeAd
+         didRenderInView:(UIView *)view
+          viewController:(UIViewController *)viewController {
+    NSMutableArray *assets = [[NSMutableArray alloc] init];
+    if ([view isKindOfClass:[GADNativeAppInstallAdView class]]) {
+        GADNativeAppInstallAdView *adView = (GADNativeAppInstallAdView *)view;
+        if (adView.headlineView != nil) {
+            [assets addObject:adView.headlineView];
+        }
+        if (adView.imageView != nil) {
+            [assets addObject:adView.imageView];
+        }
+        if (adView.iconView != nil) {
+            [assets addObject:adView.iconView];
+        }
+        if (adView.adChoicesView != nil) {
+            [assets addObject:adView.adChoicesView];
+        }
+        if (adView.bodyView != nil) {
+            [assets addObject:adView.bodyView];
+        }
+        if (adView.callToActionView != nil) {
+            [assets addObject:adView.callToActionView];
+        }
+        if (adView.priceView != nil) {
+            [assets addObject:adView.priceView];
+        }
+        if (adView.starRatingView != nil) {
+            [assets addObject:adView.starRatingView];
+        }
+        if (adView.storeView != nil) {
+            [assets addObject:adView.storeView];
+        }
+        if (adView.mediaView != nil) {
+            [assets addObject:adView.mediaView];
+        }
+    } else {
+        NSLog(@"View is not the instance of GADNativeAppInstallAdView, Failed to register View for "
+              @"user interaction");
+    }
+    
+    [_nativeAd registerViewForInteraction:view
+                       withViewController:viewController
+                       withClickableViews:assets];
+}
+
+- (void)mediatedNativeAd:(id<GADMediatedNativeAd>)mediatedNativeAd didUntrackView:(UIView *)view {
+    [_nativeAd unregisterView];
+}
+
+#pragma mark - DUNativeAdDelegate
+
+- (void)nativeAdDidLoad:(DUNativeAd *)nativeAd {
+    [self loadNativeAdImages];
+}
+
+- (void)nativeAdWillLogImpression:(DUNativeAd *)nativeAd {
+    if (_impressionLogged) {
+        GADDU_LOG(@"DUNativeAd is trying to log an impression again. Adapter will ignore duplicate "
+                  "impression pings.");
+        return;
+    }
+    
+    _impressionLogged = YES;
+    [GADMediatedNativeAdNotificationSource mediatedNativeAdDidRecordImpression:self];
+}
+
+- (void)nativeAd:(DUNativeAd *)nativeAd didFailWithError:(NSError *)error {
+    id<GADMAdNetworkConnector> strongConnector = _connector;
+    id<GADMAdNetworkAdapter> strongAdapter = _adapter;
+    [strongConnector adapter:strongAdapter didFailAd:error];
+}
+
+- (void)nativeAdDidClick:(DUNativeAd *)nativeAd {
+    [GADMediatedNativeAdNotificationSource mediatedNativeAdDidRecordClick:self];
```

Comment:
Since a click would lead to an user leaving the application, you can send the `adapterWillLeaveApplication` callback here as.

Reply:
I use StoreKit to show the AD details in apps, in this case, should I send the adapterWillLeaveApplication callback? (interstitial AD in the same situation)